activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -4,16 +4,52 @@ module ActiveRecord
   module Associations
     class Preloader
       class Association #:nodoc:
-        def self.load_records_in_batch(scope, association_key_name, loaders)
-          ids = loaders.flat_map(&:owner_keys).uniq
+        class LoaderQuery
+          attr_reader :scope, :association_key_name
 
-          raw_records = scope.where(association_key_name => ids).load do |record|
-            loaders.each { |l| l.set_inverse(record) }
+          def initialize(scope, association_key_name)
+            @scope = scope
+            @association_key_name = association_key_name
           end
 
-          loaders.each do |loader|
-            loader.load_records(raw_records)
-            loader.run
+          def eql?(other)
+            return false unless association_key_name == other.association_key_name
+
+            # Attempt to inexpensively rule out or confirm equality
+            return false unless scope.table_name == other.scope.table_name
+            return false unless scope.preload_values == other.scope.preload_values
+            return false unless scope.includes_values == other.scope.includes_values
+            return true if scope.empty_scope? && other.scope.empty_scope?
+            return false unless scope.where_values_hash == other.scope.where_values_hash
```

Comment:
We can use `scope.where_clause` here.

```suggestion
            return false unless scope.where_clause == other.scope.where_clause
```